cpp/src/centrality/betweenness_centrality.cuh

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Author: Xavier Cadet xcadet@nvidia.com
+#pragma once
+
+namespace cugraph {
+namespace detail {
+template <typename VT, typename ET, typename WT, typename result_t>
+class BC {
+   private:
+      // --- Information concerning the graph ---
+      const experimental::GraphCSR<VT, ET, WT> &graph;
+      // --- These information are extracted on setup ---
+      VT number_of_vertices;  // Number of vertices in the graph
+      VT number_of_edges;     // Number of edges in the graph
+      ET const *offsets_ptr;  // Pointer to the offsets
+      VT const *indices_ptr;  // Pointers to the indices
+
+
+      // --- Information from configuration ---
+      bool configured = false;         // Flag to ensure configuration was called
+      bool normalized = false;          // If True normalize the betweenness
+      // TODO: For weighted version
+      WT const *edge_weights_ptr = nullptr; // Pointer to the weights
+      bool endpoints = false;          // If True normalize the betweenness
+      VT const *sources = nullptr;     // Subset of vertices to gather information from
+      VT number_of_sources;            // Number of vertices in sources
+
+      // --- Output ----
+      // betweenness is set/read by users - using Vectors
+      result_t *betweenness = nullptr;
+
+      // --- Data required to perform computation ----
+      VT *distances = nullptr;         // array<VT>(|V|) stores the distances gathered by the latest SSSP
+      VT *predecessors = nullptr;      // array<WT>(|V|) stores the predecessors of the latest SSSP
+      double *sp_counters = nullptr;   // array<VT>(|V|) stores the shortest path counter for the latest SSSP
```

Comment:
The `sp_counters` is used to gather the number of shortest path in each vertices. It could be reused by other algorithms that rely on shortest path counting, I wanted to use `sp` for `shortest path`, the BC specific values are `betweenness` to store the result itself and `deltas` for the dependencies in the accumulation phase.